hazelcast/src/main/java/com/hazelcast/sql/impl/state/QueryClientStateRegistry.java

```diff
@@ -94,24 +94,27 @@ private SqlPage fetchInternal(
     }
 
     private static boolean fetchPage(
-        Iterator<SqlRow> iterator,
+        ResultIterator<SqlRow> iterator,
         List<List<Data>> page,
         int cursorBufferSize,
         InternalSerializationService serializationService
     ) {
-        while (iterator.hasNext()) {
+        assert cursorBufferSize > 0;
+
+        if (!iterator.hasNext()) {
+            return true;
+        }
+
+        int res;
+        do {
             SqlRow row = iterator.next();
             List<Data> convertedRow = convertRow(row, serializationService);
 
             page.add(convertedRow);
+            res = iterator.hasNextImmediately();
```

Comment:
I do not quite understand how it affects IMDG. Consider that the batch is not available immediately for IMDG. Will it stop waiting for results and return a semi-filled batch back to the client?

Reply:
Yes, if `waitForFullBatch` is false. For IMDG queries, it's always `true`, in this case this call will block until next item is availabye. The return value will never be `RETRY`.

Reply:
Got it, thanks.